File: battleship/fsm.py
""" Very basic finite state machinery """

class State(object):
	__slots__ = ('on_enter', 'on_exit', 'events')
	def __init__(self, on_enter=None, on_exit=None, events=None):
		self.on_enter = on_enter
		self.on_exit = on_exit
		self.events = events

class Event(object):
	__slots__ = ('input', 'transitions')
	def __init__(self, input=None, transitions=None):
		self.input = input
		self.transitions = transitions

class Transition(object):
	__slots__ = ('condition', 'state', 'action')
	def __init__(self, condition=None, state=None, action=None):
		self.condition = condition
		self.state = state
		self.action = action

class Transducer:
	def __init__(self, states, initial_state, context):
		self.context = context
		self.states = states
		self.__set_state(initial_state)

	def __set_state(self, state_id):
		st = self.states[state_id]
		if st.on_enter:
			st.on_enter(self.context)
		self.current_state_id = state_id
		return state_id

	def dispatch(self, event_id, input_data):
		st = self.states.get(self.current_state_id)
		assert st, "STATE %s" % self.current_state_id
		event = st.events.get(event_id)
		if event is None:
			raise Exception("STATE %s have no EVENT %s" % (self.current_state_id, event_id))
		event.input(self.context, input_data)
		for tr in event.transitions:
			if tr.condition and tr.condition(self.context):
				if tr.action:
					tr.action(self.context)
				self.__set_state(tr.state)
				return

	def get_state(self):
		return self.current_state_id

File: battleship/game/fsm.cpp
#include "fsm.h"
#include <exception>
#include <assert.h>

namespace fsm {


template <typename CTX, typename DATA>
Transducer<CTX,DATA>::Transducer(const std::map<int, State<CTX,DATA> >& states, int initial_state, CTX* context)
{
	p_context = context;
	m_states = states;
	SetState(initial_state);
}


template <typename CTX, typename DATA>
void Transducer<CTX,DATA>::Dispatch(int event_id, DATA& input) throw (InvalidEventException)
{
	std::map<int, State<CTX,DATA> >::iterator st = m_states.find(m_state);
	assert (st != m_states.end());
	std::map<int, Event<CTX,DATA> >::iterator ev = st->second.events.find(event_id);
	if (ev == st->second.events.end())
		throw InvalidEventException();
	ev->second.input(p_context, input);
	std::vector<Transition<CTX,DATA> >::iterator tr;
	for (tr = ev->second.transitions.begin(); tr != ev->second.transitions.end(); tr++)
	{
		if (tr->condition && tr->condition(p_context))
		{
			if (tr->action)
				tr->action(p_context);
			SetState(tr->state);
			return;
		}
	}
}


template <typename CTX, typename DATA>
int Transducer<CTX,DATA>::GetState()
{
	return m_state;
}


template <typename CTX, typename DATA>
int Transducer<CTX,DATA>::SetState(int state)
{
	State<CTX,DATA>& st = m_states[state];
	if (st.onEnter)
		st.onEnter(p_context);
	m_state = state;
	return state;
}




}  // namespace
